igvc_navigation/src/navigation_client/navigation_client.cpp

```diff
@@ -0,0 +1,184 @@
+{
+  ros::NodeHandle private_nh("~");
+
+  assertions::getParam(private_nh, "reading_from_file", reading_from_file_);
+  assertions::getParam(private_nh, "/waypoint_file_path", waypoint_file_path_);
+  assertions::getParam(private_nh, "waypoint_radius", waypoint_radius_);
+
+  rviz_sub_ = nh_.subscribe("/move_base_simple/goal", 1, &NavigationClient::rvizWaypointCallback, this);
+
+  waitForTransform();
+  waitForServer();
+
+  if (reading_from_file_)
+  {
+    loadWaypointsFile();
+    for (const geometry_msgs::PointStamped& waypoint : waypoints_list_)
+    {
+      sendPointAsGoal(waypoint, true);
+    }
+  }
+  else
+  {
+    ROS_INFO_STREAM("Waiting for waypoints from rviz.");
+  }
+}
+
+void NavigationClient::waitForTransform()
+{
+  const double waiting_time = 5.0;
+  while (!tf_listener_.waitForTransform("odom", "utm", ros::Time(0), ros::Duration(waiting_time)))
+  {
+    ROS_INFO_STREAM("utm->odom transform not found. waiting...");
+  }
+  ROS_INFO_STREAM("utm->odom transform found!");
+}
+
+void NavigationClient::waitForServer()
+{
+  const double waiting_time = 5.0;
+  while (!client.waitForServer(ros::Duration(waiting_time)))
+  {
+    ROS_INFO("Waiting for the move_base action server to come up");
+  }
+}
+
+void NavigationClient::loadWaypointsFile()
+{
+  ROS_INFO_STREAM_ONCE("Loading waypoints from " << waypoint_file_path_);
+
+  if (waypoint_file_path_.empty())
+  {
+    ROS_ERROR_STREAM("Could not load waypoints. Empty file path.");
+    ros::shutdown();
+  }
+
+  std::ifstream file;
+  file.open(waypoint_file_path_.c_str());
+
+  if (!file.is_open())
+  {
+    ROS_INFO_STREAM("Could not open file: " << waypoint_file_path_);
+    ros::shutdown();
+  }
+
+  std::string line;
+  auto line_index = 1;
+  while (!file.eof())
+  {
+    getline(file, line);
+
+    if (!line.empty() && line[0] != '#')
+    {
+      std::vector<std::string> tokens = split(line, ',');
+
+      if (tokens.size() != 2)
+      {
+        ROS_ERROR_STREAM(waypoint_file_path_ << ":" << line_index << " - " << tokens.size() << " tokens instead of 2.");
+        return;
+      }
+
+      // convert latitude and longiture to decimal degrees if currently in degrees minutes second. specified by '?'
+      // symbol
+      double lat = (tokens[0].find('?') != std::string::npos) ? convertDmsToDec(tokens[0]) : stod(tokens[0]);
+      double lon = (tokens[1].find('?') != std::string::npos) ? convertDmsToDec(tokens[1]) : stod(tokens[1]);
+
+      // transform latitude and longitude to UTM frame
+      geometry_msgs::PointStamped waypoint_utm;
+      RobotLocalization::NavsatConversions::UTM(lat, lon, &(waypoint_utm.point.x), &(waypoint_utm.point.y));
+      waypoint_utm.header.frame_id = "utm";
+
+      // transform utm frame to odom frame
+      geometry_msgs::PointStamped waypoint_odom;
+      tf_listener_.transformPoint("odom", ros::Time(0), waypoint_utm, "odom", waypoint_odom);
+      waypoints_list_.push_back(waypoint_odom);
+    }
+    line_index++;
+  }
```

Comment:
And I still think this section could be a bit more cleaner, maybe something like
```c++
const std::vector<LatLong> latlongs = parseFile(waypoint_file_path_);
for (cons auto& latlong : latlongs)
{
 waypoints_list_.emplace_back(latlongToPoint(latlong));
}
```

also, you can clean up the `getline` stuff by doing
```c++
while (std::string line; std::getline(file, line))
{
  ...
}
```

or something